types/occ/scheduler.go

```diff
@@ -18,3 +23,147 @@ func NewEstimateAbort(dependentTxIdx int) Abort {
 		Err:            ErrReadEstimate,
 	}
 }
+
+type TaskStatus string
+
+const (
+	TaskStatusPending   TaskStatus = "pending"
+	TaskStatusExecuted  TaskStatus = "executed"
+	TaskStatusAborted   TaskStatus = "aborted"
+	TaskStatusValidated TaskStatus = "validated"
+)
+
+type Task struct {
+	Status      TaskStatus
+	Index       int
+	Incarnation int
+	Request     types.RequestDeliverTx
+	Response    *types.ResponseDeliverTx
+}
+
+// TODO: (TBD) this might not be necessary to externalize, unless others
+func NewTask(request types.RequestDeliverTx, index int) *Task {
+	return &Task{
+		Request: request,
+		Index:   index,
+		Status:  TaskStatusPending,
+	}
+}
+
+// Scheduler processes tasks concurrently
+type Scheduler interface {
+	ProcessAll(ctx sdk.Context, tasks []*Task) error
+}
+
+type scheduler struct {
+	deliverTx func(ctx sdk.Context, req types.RequestDeliverTx) (res types.ResponseDeliverTx)
+	workers   int
+}
+
+// NewScheduler creates a new scheduler
+func NewScheduler(workers int, deliverTxFunc func(ctx sdk.Context, req types.RequestDeliverTx) (res types.ResponseDeliverTx)) Scheduler {
+	return &scheduler{
+		workers:   workers,
+		deliverTx: deliverTxFunc,
+	}
+}
+
+func (s *scheduler) ProcessAll(ctx sdk.Context, tasks []*Task) error {
+	toExecute := tasks
+	for len(toExecute) > 0 {
+
+		// execute sets statuses of tasks to either executed or aborted
+		err := s.executeAll(ctx, toExecute)
+		if err != nil {
+			return err
+		}
+
+		// validate returns any that should be re-executed
+		// note this processes ALL tasks, not just those recently executed
+		toExecute, err = s.validateAll(ctx, tasks)
```

Comment:
we want validation to also be ongoing and not done after all executions are done, that way if we fail validation for tx X we can immediately re-execute instead of having waited for ALL the txs to have executed before realizing the validation failed

Reply:
Is it clear that as soon as validation fails is when we know it should be immediately re-executed?  (If we have to prioritize re-executions?)  That wasn't yet clear to me because it seems like we have to do a full-validation at the end anyway.